### ISSUE TITLE
C++并发编程实战_笔记

### ISSUE THREAD

**Issue by v4if** [v4if/blog#12]
### 启动线程
`std::thread my_thread(background_task());`
这里相当于声明了一个名为my_thread的函数，这个函数带有一个参数（函数指针指向没有参数并返回background_task对象的函数），返回一个std::thread对象的函数，而非启动了一个线程

```c++
// 使用多组括号
std::thread my_thread((background_task()));

// 或使用新统一的初始化语法
std::thread my_thread{background_task()};

// 使用lambda表达式，允许使用一个可以捕获局部变量的局部函数
std::thread my_thread([]{
  do_something();
});
```
启动了线程，需要明确是要等待线程结束(join)，还是让其自主运行(detach)

### 悬空引用
如果不等待线程，就必须保证线程结束之前，可访问的数据的有效性
这种情况和可能发生在线程还没结束，函数已经退出的时候，这时线程函数还持有函数局部变量的指针或引用
```c++
struct func {
    int& i_;
    func(int& i) : i_(i){}
    void operator()() {
        using namespace std;
        for (unsigned int i = 0; i < 65535; ++i) {
            cout << i_ << endl;     // 潜在访问隐患：悬空引用
        }
    }
};

// oops崩溃
void oops() {
    int local_state = 0;
    func my_func(local_state);

    std::thread t1(my_func);
    t1.detach();
    /* 不等待线程结束，新线程可能还在运行
      * */
}
```

### 特殊情况下的等待
```c++
struct func {
    int& i_;
    func(int& i) : i_(i){}
    void operator()() {
        do_something(i_);
    }
};

void do() {
    int local_state = 0;
    func my_func(local_state);

    std::thread t1(my_func);
    try {
      do_something_in_current_thread();
    } catch(...) {
      t1.join(); // 异常流
      throw;
    }

    t1.join(); // 正常流
}
```
使用了try/catch块确保访问本地状态的线程退出后，线程才结束

还有一种方式是使用`资源获取即初始化方式`，并且提供一个类，在析构函数中使用join()
```c++
struct func;
class thread_guard{
    std::thread& t_;
public:
    explicit thread_guard(std::thread& t) : t_(t) {}
    ~thread_guard() {
        if (t_.joinable())
            t_.join();
    }

    thread_guard(const thread_guard&) = delete;
    thread_guard& operator=(const thread_guard&) = delete;
};

void do() {
    int local_state = 0;
    func my_func(local_state);

    std::thread t1(my_func);
    thread_guard tg(t1);

    do_something_in_current_thread();
}
```
join()只能对给定的对象调用一次，所以对给已加入的线程再次进行加入操作时，将会导致错误

如果不想等待线程结束，可以分离(detach)线程，从而避免异常安全问题。不过，这就打破了线程与std::thread对象的联系，即使线程仍然在后台运行着，分离操作也能确保std::terminate()在std::thread对象销毁才被调用

### 后台线程
使用detach()会让线程在后台运行，这就意味着主线程不能与之产生直接交互。也就是说，不会等待这个线程结束；如果线程分离，那么就不可能有std::thread对象能引用它，分离线程的确在后台运行，所以分离线程不能被加入。不过C++运行库保证，线程退出时，相关资源能够正确回收，后台线程的归属和控制C++运行库都会处理。通常称分离线程为守护线程

### 向线程函数传递参数
需要注意的是，默认参数要拷贝到线程独立内存中，即使参数是引用的形式，也可以在新线程中进行访问
```c++
void f(int i, std::string const& s);  
void oops(int some_param) {
  char buffer[1024];
  sprintf(buffer, "%i", some_param);
  std::thread t(f, 3, buffer);
  t.detach();
}
```
这种情况下，buffer是一个指针变量，指向本地变量，然后本地变量通过buffer传递到新线程中。并且函数有很大可能，会在字面值转化成std::string对象之前崩溃，从而捯饬线程的一些未定义行为。解决方案是在传递到std::thread构造函数之前就将字面值转化为std::string对象
`std::thread t(f, 3, std::string(buffer));`

`std::bind`
bind的思想实际上是一种延迟计算的思想，将可调用对象保存起来，然后在需要的时候再调用
```c++
void add(int a, int b, int& r) {
    r = a + b;
}
int main() {
    int ret = 0;
    // 函数柯里化，_1相当于占位符
    auto f = std::bind(add, std::placeholders::_1, 20, std::ref(ret));
    f(10);
    std::cout << ret << std::endl; // 30
    return 0;
}
```
std::ref返回的是std::reference_wrapper，是一个引用封装类
假如`std::bind(add, std::placeholders::_1, 20, ret)`传递给函数的参数是ret变量内部拷贝的引用，而非数据本身的引用

```c++
class X{
  public:
    void do_lengthy_work();
};
X my_x;
std::thread t(&X::do_lengthy_work(), &my_x);
```
可以传递一个成员函数指针作为线程函数，并提供一个合适的对象指针作为第一个参数

提供的参数可以`移动(move)`，但不能`拷贝(copy)`。移动是指：原始对象中的数据转移给另一对象，而转移的这些数据就不再在原始对象中保存了。使用`移动`转移原对象后，就会留下一个空指针(NULL)
展示了std::move()是如何转移一个动态对象到一个线程中去的
```c++
void process_big_object(std::unique_ptr<big_object>);

std::unique_ptr<big_object> p(new big_object);
p->prepare_data(42);
std::thread t(process_big_object, std::move(p));
```
在std::thread的构造函数中指定std::move(p)，big_object对象的所有权就被首先转移到新创建线程的内部存储中，之后传递给process_big_object函数

### 转移线程所有权
在同一时间点，就能保证只关联一个执行线程
`当原对象是一个临时变量时，自动进行移动操作`
```c++
void some_function();
void some_other_function();

std::thread t1(some_function);
std::thread t2 = std::move(t1);

t1 = std::thread(some_other_function); // 右值，即将消亡  临时对象

std::thread t3;
t3 = std::move(t2);

t1 = std::move(t3); //赋值操作将使程序崩溃
```
t1已经有了一个关联的线程(执行some_other_function的线程)，所以这里直接调用std::terminate()终止程序继续执行。终止操作将调用std::thread的析构函数，销毁所有对象(与C++中异常的处理方式很相似)

当所有权可以在函数内部传递，就允许std::thread实例可以作为参数进行传递
```c++
void f(std::thread t);
void g() {
    void some_function();
    f(std::thread(some_function));

    std::thread t(some_function);
    f(std::move(t));
}
```
std::thread支持移动的好处是可以创建thread_guard类的实例，并且拥有其线程的所有权
为了确保线程程序在退出前完成，可以定义scoped_thread类
```c++
class scoped_thread{
    std::thread t_;
    public:
        explicit scoped_thread(std::thread t) : t_(std::move(t)) {
            if(!t_.joinable) 
                throw std::logic_error("No thread");
        }

        ~scoped_thread() {
            t_.join();
        }

        scoped_thread(scoped_thread const&) = delete;
        scoped_thread& operator=(scoped_thread const&) = delete;
};

struct func;
void f() {
    int some_local_state;
    // 新线程直接传递到scoped_thread中，而非创建一个独立的命名变量
    scoped_thread t(std::thread(func(some_local_state))); 
    do_something_in_current_thread();
}
```

`量产线程`
```c++
void do_work(unsigned id);
void f() {
    std::vector<std::thread> threads;
    for (unsigned i = 0; i < 20; i++) {
        threads.push_back(std::thread(do_work, i));
    }

// 对每个线程调用join
    std::for_each(threads.begin(), threads.end(), 
        std::mem_fn(&std::thread::join));
}
```

### 运行时决定线程数量
